### DISCUSSION TITLE
Addon Project: The Styles Palette

### DISCUSSION THREAD

**Discussion by euu2021** [freeplane/freeplane#415] [Show and tell]
I'm starting a project with the objective of creating a script/addon that helps in the use of styles.

I feel like I'm underusing styles and I would use them more if it was easier to preview the styles when choosing them. I explained it in detail [here](https://sourceforge.net/p/freeplane/discussion/758437/thread/b01f757363). So, the gold standard that I take for the start is the Scapple (iOS version) behavior.

I plan to do it in progressively better implementations, as my scripting skill level improves (yes, it will take a long time). Right now, what my skill level allows me to create is this:

An Autohotkey script that allows to, with one hotkey, open the Style Manager and then apply the selected style. So, I simply select one or more nodes:
![image](https://user-images.githubusercontent.com/77707706/170888322-2dfeff9a-8529-4bbc-905f-cae98b5c6f16.png)

Press the hotkey, and it opens the Style Manager. I select the desired style:
![image](https://user-images.githubusercontent.com/77707706/170888351-1933b524-7234-4243-996a-7fcae2add9d2.png)

Press the same hotkey, the Style Manager is closed, and the style applied to the node(s):
![image](https://user-images.githubusercontent.com/77707706/170888364-53d9856e-f2ca-4800-b88a-1ab84c199375.png)

It's simple, but it already makes the experience more comfortable.

The script is this:

```
#NoEnv  ; Recommended for performance and compatibility with future AutoHotkey releases.
; #Warn  ; Enable warnings to assist with detecting common errors.
SendMode Input  ; Recommended for new scripts due to its superior speed and reliability.
SetWorkingDir %A_ScriptDir%  ; Ensures a consistent starting directory.
SetTitleMatchMode 2

^j::
if WinActive("Freeplane - Style editing")
{
Send, !+C

sleep, 200

Send, {Esc}

sleep, 200

Send, !+v
return
}

else if WinActive("ahk_exe javaw.exe")
{
Send, ^{F11}
return
}

else
```
Change `^j`  (Ctrl+J) for the hotkey you want to be used to activate the script.
Change `^{F11}` (Ctrl+F11) for the hotkey that you use to open the Style Manager.

So this first implementation is through an Autohotkey script, because I find it much easier than creating FP groovy scripts. Please, share some ideas about how to make it evolve to the next level. I will put my wishlist in the end of this post. I was thinking about going in the path of creating a FP script that does something like that:

1. Create a new node X in the root or floating anywhere that doesn't get in the way (maybe on a new map?)
2. Record the selected nodes ids as a property of node X, so I can point to those nodes as the ones for which the style actions will be applied
3. Create new children in node X, and apply each style available, so I end up with a list of all styles
4. Open a second map view, docked on the side (or simply another window), centered on the node X, and filtered to show only it
5. Open a floating menu with styling actions to be applied to the nodes listed in the properties of node x
6. Close the second map view and delete node X, when the action is applied

My wishlist (please, suggest more ideas): 

- [x] Opening a list of styles with preview, and easily applying a style chosen from it
- [ ]     Inserting nodes from the Style Palette
- [ ]     Defining and redefining styles via drag and drop to the style palette
- [ ]     Grouping styles, like Edo suggested [here](https://sourceforge.net/p/freeplane/discussion/758437/thread/2e01393ab5/?limit=25#2fb0).
- [ ]     Showing only user defined styles
- [ ]     Checkboxes to show only certain groups of styles
- [ ]     Using the Style Palette for easier filtering and highlighting of nodes
- [ ]     From the palette, define a default node style (i.e., the style to be used on any new node)

BR,

**Comment by rexelbartolome** [comment]
Ohhh that's interesting. I haven't thought about the map styles that way. However, I do think that the current set of styles are a bit too similar with each other and was thinking of designing some new ones with better font pairings and more variety overall. For example, Freeplane currently doesn't offer a 'handwritten' map style by default while other mindmapping programs usually offer at least one of those styles by default.

I also think that the closer we can get to a 'live preview' of a style, the better. So the AHK script you made definitely made that gap closer to a 'live preview' since other ways of applying a style is through 

- a keyboard shortcut
- right-click node >`Apply Style`
- `Format > Apply Style`

and none of them has a preview of the style being applied. Regardless, very excited to what this addon can achieve :D 

Speaking of AHK scripts, I'm absolutely in love with this one https://github.com/dumbeau/AutoHotPie
It gives a radial pie menu which you can map keyboard shortcuts, scripts, and other macros to. Not really relevant to your addon or Freeplane, just wanted to share how great it is 😆

**Reply by euu2021** [reply]
Yes, the templates are currently poor in styles variety.

I didn't know AutoHotPie. It seems to be very useful. Thanks for the suggestion.

**Comment by euu2021** [comment]
I improved the script a bit, in order to make the mouse automatically jump to styles list and back to the selected node. It saves the mouse coordinates in an ini file, which allows to remember the last position in the styles list, making it easy to apply again the last applied style.
![gjInVL7odb](https://user-images.githubusercontent.com/77707706/171182154-a39c14a0-a7ed-4784-a538-e6f572291cab.gif)

But... I noticed that I'm copying only the format, and not the actual style 🤦‍♂️. 

I don't know how to copy the style from Styles Manager. I think one solution would be to write the name of the style in the details of the  node, and then make the script copy that name, so it can be used as the key to setStyle.

The problem is that I don't know how to use setStyle. So, my question is, if I have a style named Style1, how do I set this style to a node using a Freeplane script? I tried things like that, with no success:
`node.style.setStyle('Style1')`

**Reply by abc16361** [reply]
`node.style.setName("Style1")`  or `node.style.name = "Style1"`

**Reply by euu2021** [reply]
Thanks!

**Comment by euu2021** [comment]
It's not possible to copy the core content or name of the style in the Style Editor, so the solution that I found is to copy the details of the node in the Style Editor (so, the user needs, previously, to write the name of the style one by one), and activate a FP script that reads the name in the clipboard and set the node style to that.

It's ugly, but it works:
![javaw_6AeO7We6bW](https://user-images.githubusercontent.com/77707706/171278788-c04424cb-2e1d-419e-84f5-eb35ecd58ec3.gif)


Here is the Autohotkey side:
```
#NoEnv 
SendMode Input 
SetWorkingDir %A_ScriptDir% 
SetTitleMatchMode 2
CoordMode, Mouse

/*
Things to change, acording to your needs:

-Hotkey to use the script = MButton (i.e., Middle mouse button)
-Hotkey used in FreePlane to open the Style Editor = Ctrl+F11
-Hotkey used in FreePlane to edit node details inside the Style Editor = ^{F3}
-Hotkey to activate the auxiliary FP script = ^{4}
-Folder where to create the ini file, that stores variables: C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini
*/

#IfWinActive, Freeplane - Style editing
MButton::

MouseGetPos x, y
IniWrite, %x%, C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini, onPalette, x
IniWrite, %y%, C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini, onPalette, y
Send, ^{F3}
sleep, 100
Send, ^{a}
sleep, 100
Send, ^{c}
sleep, 100
Send, {Enter}
sleep, 100
Send, ^{Enter}
sleep, 100
Send, ^{4}
IniRead, mx, C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini, onMap, x
IniRead, my, C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini, onMap, y
MouseMove %mx%, %my%
MouseMove, 01, 01, 01, R
MouseMove, -01, -01, 01, R

return

#IfWinActive, ahk_exe javaw.exe
MButton::

MouseGetPos x, y
IniWrite, %x%, C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini, onMap, x
IniWrite, %y%, C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini, onMap, y


Send, ^{F11}

IniRead, px, C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini, onPalette, x
IniRead, py, C:\Users\kauen\Documents\ferramentas\Autohotkey\RunScript_Notepad++\myfile.ini, onPalette, y
MouseMove, %px%, %py%
MouseMove, 01, 01, 01, R
MouseMove, -01, -01, 01, R

return

```

And the FP side (it's from [here](https://sourceforge.net/p/freeplane/discussion/758437/thread/0c97a52260)):
````
import java.awt.datatransfer.*
import java.awt.Toolkit

Clipboard clipboard = Toolkit.getDefaultToolkit().getSystemClipboard()

clipboardContents = clipboard.getContents(null).getTransferData(DataFlavor.stringFlavor)

styleToSet = clipboardContents

node.style.setName(styleToSet)
````

**Reply by rexelbartolome** [reply]
Hmmm that's too bad, I'm not sure I understand how you got to put `details` inside the Style Editor? 

Either way, maybe a more elegant solution would be somehow exporting the Styles Editor mindmap into an actual .mm file. That way, you'll be in complete control of the map and it's also editable. It could also come out as a side panel or a floating window just like the Styles Editor. It could also pave the way for you to categorize the `user defined styles` and omit `Pre-defined styles` and `level styles` since it is after all just another mind map that you can edit.

https://user-images.githubusercontent.com/88552647/171283065-ec10abf1-2e3b-4f6b-9a4b-5f19b3c3326c.mp4


https://user-images.githubusercontent.com/88552647/171283190-0d852aa3-f39a-43de-b1ea-64b88a7495e8.mp4

**Reply by euu2021** [reply]
> Hmmm that's too bad, I'm not sure I understand how you got to put details inside the Style Editor?

Inside the Style Editor, navigate to Edit, in the menubar. There you can insert details, notes, and attributes. It's useful to insert comments about how do you use the style, for example, or to give an alternative name to the style, as the process of changing the name is not convenient.

In the script above I forgot that I specified a custom hotkey for that. I will edit.

**Comment by EdoFro** [comment]
I think @rexelbartolome proposal makes more sence:
With a groovy script you can create a new map that have the same styles (related) that have nodes with all of the available node's styles.
It can also use the outline view so it stays narrow.
and with another script you can read the selected node style in one map and apply it to the selected nodes in the other.

or something like that.

jmho

**Reply by EdoFro** [reply]
get a look to this [script](https://github.com/EdoFro/Freeplane_groovyNode/blob/main/groovyNode/scripts/importGroovyNodeStyles.groovy) and the [class library](https://github.com/EdoFro/Freeplane_groovyNode/blob/main/groovyNode/src/main/groovy/UserStyles.groovy) it uses. Both are written in groovy.

I use them to copy styles from template maps to other maps.

You can use them to access the style nodes in your map and build the script you need.

I think the `getUserDefinedStylesParentNode()` method can help you. ([here](https://github.com/EdoFro/Freeplane_groovyNode/blob/433665cb92d241b6c2ca29b11fbd02c49625f9e3/groovyNode/src/main/groovy/UserStyles.groovy#L117))

hope it helps, edo

**Reply by EdoFro** [reply]
with them I can copy only the user styles without affecting the other styles

**Reply by rexelbartolome** [reply]
All the links you've posted go to a 404 page in github, might be something with permissions or the contents being a private repo instead of a public one?

**Reply by euu2021** [reply]
I think the repositories that you linked are private. Yes, I would love to learn from them, please.

**Reply by EdoFro** [reply]
sorry, yes, it was private. Now it’s public.

**Comment by euu2021** [comment]
The problem that I see in placing the palette on a second map is that it would be much harder to enable the drag and drop operations. Right now, I don't see disadvantages in creating the palette in the same map, inside a floating node formatted to be invisible. But, those methods to import styles from other map will be useful, anyway, because I want to create the option to show in the palette styles from other maps. Thanks.

I will play with withView() to see how it works. I would like to have a map view created right beside the mouse, like a tooltip, allowing a quick interaction.

**Reply by EdoFro** [reply]
If it's a second map or not, it's your choice. I think a floating node is a great idea: much easier and powerful.

I really think  the `getUserDefinedStylesParentNode()` method is what you need. It returns the parent node of the user styles nodes from the styles map. By asking for its children you have all the names of the existing  user styles of the map.

**Reply by EdoFro** [reply]
here I wrote a sample script.

I left some methods not needed for this example but also interesting (I commented them so you can easily recognize which are the used methods here)

What it does: It inserts sample nodes of all user styles as childs of the selected node

```groovy
import org.freeplane.api.MindMap as ApiMindMap

import org.freeplane.plugin.script.proxy.NodeProxy as ProxyNode
import org.freeplane.features.map.MapModel;
import org.freeplane.features.map.NodeModel;
import org.freeplane.features.mode.Controller;
import org.freeplane.features.styles.MapStyleModel;
import org.freeplane.plugin.script.ScriptContext

//import org.freeplane.core.ui.components.UITools as ui
import org.freeplane.plugin.script.proxy.ScriptUtils

def userStylesParentNode = getUserDefinedStylesParentNode()
def styleNames = userStylesParentNode.children*.text

def stylesSamplerParentNode = node

styleNames.each{s ->
    stylesSamplerParentNode.createChild(s).style.name = s
}

//region: getting an UserStyleNode as ProxyNode from active map

// public
    // not used in this example, but are also useful methods
    // region: not used
    /*
        def static getUserStyleNode( String userStyle ){
            return getUserStyleNode( null, userStyle )
        }
    
        def static getUserStyleNode(ApiMindMap mapaProxy, String userStyle ){
            return getUserDefinedStylesParentNode(mapaProxy).children.find{it.text == userStyle}
        }
    */
    // end region: not used


    def static getUserDefinedStylesParentNode(x = null){
        return getUserDefinedStylesParentNode((ScriptContext) null)
    }

    //not used in this example. Useful if you want to get the User styles parent node from another map (not the active one)
    // region: not used
    /* 
    def static getUserDefinedStylesParentNode(MapModel mapa){
        return getUserDefinedStylesParentNode(mapa, null)
    }

    def static getUserDefinedStylesParentNode(ApiMindMap mapaProxy){
        return getUserDefinedStylesParentNode(mapaProxy.delegate, null)
    }
    */
    // end region: not used


    def static getUserDefinedStylesParentNode(ScriptContext scriptContext){
        MapModel mapa = Controller.getCurrentController().getMap();
        return getUserDefinedStylesParentNode(mapa, scriptContext)
    }
    
    def static getUserDefinedStylesParentNode(ApiMindMap mapaProxy, ScriptContext scriptContext){
        return getUserDefinedStylesParentNode(mapaProxy.delegate, scriptContext)
    }

    def static getUserDefinedStylesParentNode(MapModel mapa, ScriptContext scriptContext){
        if(!mapa) {
            return getUserDefinedStylesParentNode(scriptContext)
        }
        MapStyleModel styleModel = MapStyleModel.getExtension(mapa);
        MapModel styleMap = styleModel.getStyleMap();
        NodeModel userStyleParentNode = styleModel.getStyleNodeGroup(styleMap, MapStyleModel.STYLES_USER_DEFINED);
        def userDefinedParentNode = new ProxyNode(userStyleParentNode, scriptContext)
        return userDefinedParentNode
    }
    
//end:

```

the original styles that come with some of the Freeplane templates are defined with its `Localized_Text`, which is not accessible with the API. For example, the "**Important**" styles name is `styles.important`. So this script doesn't work so well for those styles.

bye,
edo

**Reply by euu2021** [reply]
This is a test for a way to index scripts in a scripts collection.
---
scriptsCollection: "inserts sample nodes of all user styles as childs of the selected node"

**Comment by EdoFro** [comment]
If you want to make an add-on or a script that displays a dialog with buttons that look like the styles, you can get the style's attributes values and use them to format the button.
You can also get the css style sheet from the node. (look this method [here ](https://github.com/EdoFro/Freeplane_MarkdownHelper/blob/f6dd5a81df91eacb6595ce66cd73ad439a74ac77/MarkdownHelper/src/main/groovy/EditorStyle.groovy#L51) as an example).
I use it to style an editor dialog by defining a style in the map. This way the dialog can be customized with each mindmap.

**Reply by EdoFro** [reply]
In my opinion, some of these methods could by added to the Freeplane API. I want to test them a little more for different use cases and eventually I will propose to include them as part of the API. I think they can be used for multiple use cases.

i'm talking about the methods in [UserStyles.groovy](https://github.com/EdoFro/Freeplane_groovyNode/blob/main/groovyNode/src/main/groovy/UserStyles.groovy):

- copyUserStyles
  - it copies userstyles from one map to another, but you can define a filter criteria which styles to copy.
  - it also copies attributes and icons
  - [here](https://github.com/EdoFro/Freeplane_groovyNode/blob/433665cb92d241b6c2ca29b11fbd02c49625f9e3/groovyNode/scripts/importGroovyNodeStyles.groovy#L13-L27) you can see some use case examples
- getUserStyleNode
  - returns the node
- getUserDefinedStylesParentNode
  - returns the node

and in [EditorStyle.groovy](https://github.com/EdoFro/Freeplane_MarkdownHelper/blob/main/MarkdownHelper/src/main/groovy/EditorStyle.groovy):

- getStyleSheetFromNode(node)
  - returns a css stylesheet from a node (Any node, not just a style node)
  - useful to apply format to editor panes in a dialog by defining the style in a mindmap's node
- updateFormat(editorPane, node)
  - formats an editor pane to look just like the specified mindmap's node

hope it helps,
edo

**Comment by euu2021** [comment]
Edo,

-About  your script that creates a list of style nodes: it works straight away, even with original styles. It solved that part of the puzzle  ✅ Nice!

-About using a dialog with buttons that look like the styles: I contemplated that idea, but I think it will restrict the possibilities, mainly in terms of drag and drop operations.

So, now I can build a list of styles. The question is how to show the list in a convenient way.

I couldn't find any script/addon that uses undocked map views. And also nothing in the API that sounds like that. This, e.g., only focus the map view that is already opened: `c.mapLoader('C:/Users/kauen/Downloads/testmap.mm').withView().load()`. I found [this](https://www.freeplane.org/wiki/index.php/Scripts_collection#class_FreeplaneWindow), that maybe will be useful to adjust the window after it is undocked. 

I would be able to find workaroundish ways if I knew how to send keystrokes to FP via groovy script. I tried things with `sendKeys()`, with no success. Also, no examples found in the forums. Is it possible to do?

Something I realized today is that, when I learn to create those undocked map views, I will also be able to implement a feature that I've seen in Ayoa, and I think is brilliant: idea bank. It's like a temporary place to keep nodes that you don't want to place yet. It's shown [here](https://www.youtube.com/watch?v=XKZJJjR_5O4).

I will leave on a travel this weekend, without my laptop, and come back only in July. So, I won't be able to dive deeper into it, for now.

Thanks!

**Reply by abc16361** [reply]
Recently I have adopted an approach to put some of my scripts on a template. Everytime when I need to use these scripts, I just import this template, and close it when work is done, something like a floating tool. Speaking the styles, I think you could create a template containg nodes which represent the styles you need to use. On these nodes, add hyperlink to menu item for applying styles, and add attributes to the nodes to set conditional styles so that these nodes have the same styles as those on the style manager. Since the template is set free and have transparent edge, it can be moved anywhere you like. It can group, reorganize and fold style nodes, so it does not occupy much room. One disadvantage of this way is that it needs to create many conditional styles if many styles are going to be used.

The code to import a template: 

```
def lnk = "file:///(template-path)/template.mm" // template's url 

if (c.selected.link.text !== null) {
	def sLnk = c.selected.link.text
	c.selected.link.text = lnk
	menuUtils.executeMenuItems(['ImportLinkedBranchWithoutRootAction'])
	c.selected.link.text = sLnk
}
else {
	c.selected.link.text = lnk
	menuUtils.executeMenuItems(['ImportLinkedBranchWithoutRootAction'])
}

```
An icon tool I am testing:
![iconTool](https://user-images.githubusercontent.com/101362122/171534089-557bcea7-d85e-43e4-b6b5-c67273e32268.png)

**Reply by euu2021** [reply]
So, in your icon tool showed in the screenshot, you create a floating node and then apply the script to it? Or is it a free floating node with a hidden edge?

Your script made me remember that I can send Menu Actions from the script. This allowed me to go a couple of steps forward. Now I can imagine a sequence of actions that results in:

- A first map view apparently unchanged.
- A second map view showing only the list of styles and the root.

The basic idea is: create a Floating node; assign an invisible style (a style formatted to be transparent) to it; create the list of styles inside it; fold it, and then I have a first map view, apparently unchanged; then create a second map view; unfold the floating node; use filter selected on it, then I have a second map view with only the styles and the root.

Also, I think I know how to make the action: when the user selects the style in the list, and press the hotkey again, the styles list is deleted, the second map view is closed, and the style is applied to the nodes.

So, now the main problem is how to make this second map view be on a separate window. I will bring up the question that abc16361 made [some months ago]( https://sourceforge.net/p/freeplane/discussion/758437/thread/441f779561/): 

> Currently a linked map is opened and docked besides the working map. If we want to undock it, we need to click the undock button on the tab. I just wonder what the code is to undock the map, but cannot find the api to do it. Does anyone know the solution? Or any shortcut to undock a map?

**Reply by rexelbartolome** [reply]
I remember @dpolivaev saying that the general area where there are tabs and buttons to undock and close the mindmap are ancient parts of a deprecated library, which is part of the reason why the `tabs` of the new Look and Feel's can't be applied to that part unlike the `formatting tab`. Not sure what that implies about the API

**Reply by abc16361** [reply]
@eu2021 
my idea is to make a style template and save it to a folder. Then use the script to load it onto the working map to play as a style tool. The script also works on an undocked map.  There is no need to bother with the style manager unless you need to modify styles.
![step1a](https://user-images.githubusercontent.com/101362122/171778584-3e0104d2-fcbb-486b-aadc-181212493fb3.png)
![step2a](https://user-images.githubusercontent.com/101362122/171778603-cecc6227-54af-425c-8ae6-e9c54ced2999.png)
[ImportTemplateTest.zip](https://github.com/freeplane/freeplane/files/8829249/ImportTemplateTest.zip)

**Reply by abc16361** [reply]
Working on different maps which have different styles setup, it could need to set up different style templates. On this situation, we choose templates to load. Any better solutions?
![SelectStyleTools](https://user-images.githubusercontent.com/101362122/171798754-b964aa03-ff9b-4214-ab89-5eaaf4e760f4.gif)
[001Test.zip](https://github.com/freeplane/freeplane/files/8830049/001Test.zip)

**Reply by euu2021** [reply]
@abc16361 
Thanks for the detailed explanation. It convinced me that creating a free floating node directly in the selected node may be a better approach than dealing with undocked map views and secondary windows. I already can see that good results will come when combined with the scripts created by Edo above. Thanks!

**Reply by abc16361** [reply]
The so called selected node can be any node. Just show where the style template to import.

**Comment by euu2021** [comment]
scriptsCollection2: creates a palette of style for easy styling of nodes, with preview of the styles. Also, drag and drop directly from the palette.

I worked more on that, and now I got to a more interesting solution. A pure FP script (no need to use Autohotkey), with drag and drop:

![javaw_uqv2x2ROTG](https://user-images.githubusercontent.com/77707706/183436353-5668b6f4-ab23-45d5-bc5e-406414107d86.gif)

Use the script once and it appends the list of styles to the currently selected node. Once the list is present on the map, applying the script again will:

- If a node inside the list is selected (ie, a style node): the style will be applied to the starting node, and the list will be deleted
- if a node outside the list is selected: the list will be deleted

Here is the script (it uses a lot of the contributions by other users in the posts above):
```
import org.freeplane.api.MindMap as ApiMindMap

import org.freeplane.plugin.script.proxy.NodeProxy as ProxyNode
import org.freeplane.features.map.MapModel;
import org.freeplane.features.map.NodeModel;
import org.freeplane.features.mode.Controller;
import org.freeplane.features.styles.MapStyleModel;
import org.freeplane.plugin.script.ScriptContext

//import org.freeplane.core.ui.components.UITools as ui
import org.freeplane.plugin.script.proxy.ScriptUtils



//region: getting an UserStyleNode as ProxyNode from active map

// public
   


	def static getUserDefinedStylesParentNode(x = null){
		return getUserDefinedStylesParentNode((ScriptContext) null)
	}

	
	def static getUserDefinedStylesParentNode(ScriptContext scriptContext){
		MapModel mapa = Controller.getCurrentController().getMap();
		return getUserDefinedStylesParentNode(mapa, scriptContext)
	}
	
	def static getUserDefinedStylesParentNode(ApiMindMap mapaProxy, ScriptContext scriptContext){
		return getUserDefinedStylesParentNode(mapaProxy.delegate, scriptContext)
	}

	def static getUserDefinedStylesParentNode(MapModel mapa, ScriptContext scriptContext){
		if(!mapa) {
			return getUserDefinedStylesParentNode(scriptContext)
		}
		MapStyleModel styleModel = MapStyleModel.getExtension(mapa);
		MapModel styleMap = styleModel.getStyleMap();
		NodeModel userStyleParentNode = styleModel.getStyleNodeGroup(styleMap, MapStyleModel.STYLES_USER_DEFINED);
		def userDefinedParentNode = new ProxyNode(userStyleParentNode, scriptContext)
		return userDefinedParentNode
	}
	
if (c.findAll().collect{ it.plainText}.contains("StylesList23934223566383")) {
	if (node.parent.text == "StylesList23934223566383") {
	def styleListNode = node.parent
	def targetNode = styleListNode.parent
	def desiredStyle = node.style.name
	targetNode.style.setName(desiredStyle)
	styleListNode.delete()
	}

	else {
		c.find{
			it.text == "StylesList23934223566383"
		}.each { 
		it.delete()}
	}
}

else {
	def userStylesParentNode = getUserDefinedStylesParentNode()
	def styleNames = userStylesParentNode.children*.text

	def stylesSamplerParentNode = node.createChild()
	stylesSamplerParentNode.text = ("StylesList23934223566383")
	stylesSamplerParentNode.cloud.enabled = true
	stylesSamplerParentNode.cloud.colorCode = '#f0f0f047'
	stylesSamplerParentNode.style.setMaxNodeWidth(0)

	styleNames.each{s ->
		stylesSamplerParentNode.createChild(s).style.name = s
	}
}
```
Seems pretty functional, right now. I will improve it as I learn more about scripting.

Thanks for the help.

-
EDIT:  the script was set to find and delete a node with the exact text "Style List". So, just to be safe, I changed it to a more random thing : "StylesList23934223566383".
